api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentsServiceImpl.java

```diff
@@ -52,8 +60,28 @@ public void setAppointmentAuditDao(AppointmentAuditDao appointmentAuditDao) {
         this.appointmentAuditDao = appointmentAuditDao;
     }
 
+    private boolean validateIfUserHasSelfOrAllAppointmentsAccess(Appointment appointment) {
+        return Context.hasPrivilege(MANAGE_APPOINTMENTS) ||
+                isAppointmentNotAssignedToAnyProvider(appointment) ||
+                isCurrentUserSamePersonAsOneOfTheAppointmentProviders(appointment.getProviders());
+    }
+
+    private boolean isAppointmentNotAssignedToAnyProvider(Appointment appointment) {
```

Comment:
Minor changes required here. In multiple provider scenario, an appointment can have multiple providers, but their status might be all rejected. That needs to be considered as well. For example: "appointment.getProviders().size() == EMPTY_SET_SIZE" is not enough, since there may be providers but the provider.response might be rejected. Check "AppointmentProviderResponse".

Reply:
will review the code once.